### PR TITLE
Add `ruff` linting to pre-commit and fix all it's errors

### PR DESCRIPTION
**Pull request by savente93**
Additional linting would be good as the previous hooks didn't cover a lot. This one also adds linting for the doc strings. For some of the ones I added I didn't really have that much context, so I'll admit that they can be a bit lackluster, but I think they're okay to start off with. Feel free to improve it wherever you can. 

While I was at it I also added a short section introducing how to deal with merge conflicts with a few links for further reading.

### CODE REVIEW

File: hydromt/models/__init__.py [hunk 1]
```diff
@@ -1,17 +1,29 @@
...
+    "LumpedModel",
+    "NetworkModel",
+    "EntryPoint",
+    "ModelCatalog",
+]
 
 # NOTE: pygeos is still required in XUGRID;
 # remove requirement after https://github.com/Deltares/xugrid/issues/33
 if _compat.HAS_XUGRID and _compat.HAS_PYGEOS:
     from .model_mesh import MeshModel
```
Comment (hboisgon) [inline hydromt/models/__init__.py:23]:
```suggestion
    from .model_mesh import MeshModel, MeshMixin
```

File: hydromt/models/__init__.py [hunk 2]
```diff
@@ -1,17 +1,29 @@
...
 
+    __all__.append(MeshModel)
```
Comment (hboisgon) [inline hydromt/models/__init__.py:25]:
```suggestion
    __all__.append(MeshModel, MeshMixin)
```

File: hydromt/models/model_api.py [hunk 3]
```diff
@@ -364,13 +374,13 @@ def setup_region(
...
+        # This setup method returns region so that it can be wrapped for models which
+        # require more information, e.g. grid RasterDataArray or xy coordinates.
         return region
 
-    # TODO remove
-    # placeholder to make make sure build with the current _CLI_ARGS does not raise an error
-    def setup_basemaps(self, *args, **kwargs):
+    # TODO remove placeholder to make make sure
+    # build with the current _CLI_ARGS does not raise an error
+    def setup_basemaps(self, *args, **kwargs):  # noqa: D102
```
Comment (hboisgon) [inline hydromt/models/model_api.py:383]: What does this mean?

File: hydromt/models/model_api.py [hunk 4]
```diff
@@ -883,8 +914,10 @@ def setup_maps_from_raster_reclass(
         rename: Optional[Dict] = dict(),
         **kwargs,
     ) -> List[str]:
-        """
-        HYDROMT CORE METHOD: Add data variable(s) to maps object by reclassifying the data in ``raster_fn`` based on ``reclass_table_fn``.
+        """Add data variable(s) to maps object.
```
Comment (hboisgon) [inline hydromt/models/model_api.py:917]: Can you undo the changes here? HYDROMT CORE METHOD should be here and for the API docs online as the first line is printed the rest is needed to easily differentiate between setup_maps_from_rasterdataset and setup_maps_from_raster_reclass

File: hydromt/models/model_grid.py [hunk 5]
```diff
@@ -151,8 +154,9 @@ def setup_grid_from_raster_reclass(
         rename: Optional[Dict] = dict(),
         **kwargs,
     ) -> List[str]:
-        """
-        HYDROMT CORE METHOD: Add data variable(s) to grid object by reclassifying the data in ``raster_fn`` based on ``reclass_table_fn``.
+        """Add data variables to grid object by reclassifying the data in ``raster_fn``.
```
Comment (hboisgon) [inline hydromt/models/model_grid.py:157]: Same as for model_api, please use the same first line and undo the changes.

File: hydromt/models/model_grid.py [hunk 6]
```diff
@@ -226,35 +241,42 @@ def setup_grid_from_geodataframe(
         rename: Optional[Dict] = dict(),
         all_touched: Optional[bool] = True,
     ) -> List[str]:
-        """
-        HYDROMT CORE METHOD: Add data variable(s) to grid object by rasterizing the data from ``vector_fn``.
+        """Add data variables to grid object by rasterizing the data from ``vector_fn``.
```
Comment (hboisgon) [inline hydromt/models/model_grid.py:244]: Undo the changes

File: hydromt/models/model_grid.py [hunk 7]
```diff
@@ -291,8 +315,7 @@ def setup_grid_from_geodataframe(
 
     @property
     def grid(self):
-        """Model static gridded data. Returns xarray.Dataset.
-        Previously called staticmaps."""
+        """Model static gridded data."""
```
Comment (hboisgon) [inline hydromt/models/model_grid.py:318]:
```suggestion
        """Model static gridded data as xarray.Dataset."""
```

File: hydromt/models/model_grid.py [hunk 8]
```diff
@@ -429,11 +460,10 @@ def setup_grid(
         dec_origin: int = 0,
         dec_rotation: int = 3,
     ) -> xr.DataArray:
-        """
-        HYDROMT CORE METHOD: Create a 2D regular grid or reads an existing grid.
+        """HYDROMT CORE METHOD: Create a 2D regular grid or reads an existing grid.
 
-        An 2D regular grid will be created from a geometry (geom_fn) or bbox. If an existing
-        grid is given, then no new grid will be generated.
+        An 2D regular grid will be created from a geometry (geom_fn) or bbox. If an
```
Comment (hboisgon) [inline hydromt/models/model_grid.py:465]:
```suggestion
        A 2D regular grid will be created from a geometry (geom_fn) or bbox. If an
```

File: hydromt/models/model_mesh.py [hunk 9]
```diff
@@ -112,42 +114,64 @@ def setup_mesh_from_raster_reclass(
         rename: Optional[Dict] = dict(),
         **kwargs,
     ) -> List[str]:
-        """
-        HYDROMT CORE METHOD: Add data variable(s) to mesh object by reclassifying the data in ``raster_fn`` based on ``reclass_table_fn``.
+        """Add data variable(s) to the mesh object by reclassifying data in `raster_fn`.
```
Comment (hboisgon) [inline hydromt/models/model_mesh.py:117]: Same as grid and model, undo the changes for the first line

File: hydromt/models/model_mesh.py [hunk 10]
```diff
@@ -310,11 +341,11 @@ def setup_mesh(
         res: Optional[float] = None,
         crs: int = None,
     ) -> xu.UgridDataset:
-        """
-        HYDROMT CORE METHOD: Create an 2D unstructured mesh or reads an existing 2D mesh according UGRID conventions.
+        """Create an 2D unstructured mesh or reads an existing 2D mesh.
```
Comment (hboisgon) [inline hydromt/models/model_mesh.py:344]: Add HYDROMT CORE METHOD at the beginning

File: hydromt/models/model_network.py [hunk 11]
```diff
@@ -73,21 +77,22 @@ def write(
...
+            List of model components to write, each should have an
+            associated write_<component> method. By default ['config', 'maps',
+            'network', 'geoms', 'forcing', 'states']
         """
         super().write(components=components)
 
     # TODO: make NetworkMixin class with following properties/methods
     @property
-    def network(self):
+    def network(self):  # noqa: D102
```
Comment (hboisgon) [inline hydromt/models/model_network.py:88]: what does  # noqa: D102 mean?

File: hydromt/models/model_plugins.py [hunk 12]
```diff
@@ -93,23 +101,27 @@ def load(ep, logger=logger) -> Model:
 
 
 class ModelCatalog:
+
+    """The model cataloge provides access to plugins and their models."""
```
Comment (hboisgon) [inline hydromt/models/model_plugins.py:105]:
```suggestion
    """The model catalogue provides access to plugins and their Model classes."""
```

File: hydromt/models/model_plugins.py [hunk 13]
```diff
@@ -127,33 +139,40 @@ def plugins(self) -> List:
...
 
     def load(self, name) -> Model:
-        """Returns model class"""
+        """Return model class."""
         if name not in self._cls:
             self._cls[name] = load(self[name])
         return self._cls[name]
 
     def __str__(self):
+        """Generate string representation containing the regitered entrypoints."""
```
Comment (hboisgon) [inline hydromt/models/model_plugins.py:156]:
```suggestion
        """Generate string representation containing the registered entrypoints."""
```

File: hydromt/new_feature.py
Comment (hboisgon) [inline hydromt/new_feature.py:1]: Do we really have to commit this file? Else I would simplify the docs explanation on merging without an example

File: hydromt/raster.py [hunk 14]
```diff
@@ -26,7 +22,7 @@
 import pyproj
 import rasterio.fill
 import rasterio.warp
-import rioxarray
+import rioxarray  # noqa: F401
```
Comment (hboisgon) [inline hydromt/raster.py:25]: I'm guessing the # noqa: F401 and others are used by some of the pre-commit tools? Then good to add the most common ones and explanation in the developers docs

File: hydromt/stats/skills.py [hunk 15]
```diff
@@ -1,21 +1,17 @@
...
-import pandas as pd
 import xarray as xr
 
 warnings.filterwarnings("ignore")
 
 
 # PERFORMANCE METRICS
 def bias(sim, obs, dim="time"):
-    """Returns the bias between two time series.
+    r"""Return the bias between two time series.
```
Comment (hboisgon) [inline hydromt/stats/skills.py:14]: Why do you have some r""" here and not anywhere else:

File: hydromt/workflows/__init__.py [hunk 16]
```diff
@@ -1,7 +1,54 @@
...
 
-from .basin_mask import *
-from .forcing import *
-from .grid import *
-from .rivers import *
+from .basin_mask import get_basin_geometry, parse_region
+from .forcing import (
+    da_to_timedelta,
+    delta_freq,
+    freq_to_timedelta,
```
Comment (hboisgon) [inline hydromt/workflows/__init__.py:8]: I think from here we want pet, precip, press, temp, wind. If users need more then they can do the full import

File: hydromt/workflows/grid.py [hunk 17]
```diff
@@ -147,8 +150,9 @@ def grid_from_raster_reclass(
     mask_name: Optional[str] = "mask",
     rename: Optional[Dict] = dict(),
 ) -> xr.Dataset:
-    """
-    Prepares data variable(s) resampled to grid_like object by reclassifying the data in ``da`` based on ``reclass_table``.
+    """Prepare data variable(s) resampled to grid_like object.
```
Comment (hboisgon) [inline hydromt/workflows/grid.py:153]: Same comment as in the setup methods. the First line is what is printed in short docs api do good the full explanation is here. So would be nice to undo

File: hydromt/workflows/grid.py [hunk 18]
```diff
@@ -218,12 +226,13 @@ def grid_from_geodataframe(
     rename: Optional[Union[Dict, str]] = dict(),
     all_touched: Optional[bool] = True,
 ) -> xr.Dataset:
-    """
-    Prepares data variable(s) resampled to grid_like object by rasterizing the data from ``gdf``.
-    Several type of rasterization are possible:
-        * "fraction": the fraction of the grid cell covered by the gdf shape is returned.
-        * "area": the area of the grid cell covered by the gdf shape is returned.
-        * "value": the value from the variables columns of gdf are used. If this is used, variables must be specified.
+    """Prepare data variable(s) resampled to grid_like object.
```
Comment (hboisgon) [inline hydromt/workflows/grid.py:229]: Same here, good to keep the first line else this will be exactly the same as grid_from_raster_reclass

**Review by hboisgon — CHANGES_REQUESTED** [review]
Hi @savente93. Well this was quite some work and quite some time time to review! Nice changes also since black does not format strings and sorry you needed to add some docstrings yourself.
I have a few things that I would like to be addressed before this can be merged:

- deleted imports in _compat.py
- the * imports have been removed. Could you test it with some plugins or shall I do it?
- did you generate the docs locally to check if your changes format in the dev docs are correct?
- change some of the docstrings (see my suggestions and comments)
- add some docs on ruff comment in the developer guide
- I wonder if some of the pre-commit things should not be added to the changelog? Depending on sensibility maybe at least mention that internal * imports have been removed?

File: docs/dev/contributing.rst [hunk 19]
```diff
@@ -185,22 +275,101 @@ You can install pre-commit by running:
...
 
   $ pre-commit install
 
-Now pre-commit will check all configured hooks against all the staged files and in many cases apply fixes for you before you commit.
+After doing this pre-commit will check all your staged files when commiting.
 
-When pre-commit applies fixes for you you will have to add them to the staging area again. The git staging area will never be modified by pre-commit.
+For example say that you've added the following new feature:
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (savente93) [reply docs/dev/contributing.rst:294]: Well spotted. I think indeed black does not check rts files, and I missed this because I assumed it was picking up on the example I'd added as an actual merge conflict. Should be good now!

File: hydromt/__init__.py
[the remark this answers is not in the document]
Comment (savente93) [reply hydromt/__init__.py:1]: Good that you mention this, I'd totally forgotten about the plugins. Probably it's safest if I just revert these changes and adjust the linter.

File: hydromt/models/model_network.py (hunk 11 above)
Comment (savente93) [reply hydromt/models/model_network.py:88]: This tells the linter to ignore rule D102 (i.e. public methods need to have a doc string) for this line only. Since these are abstract methods I did not think a doc string was really necessary.

File: hydromt/stats/skills.py (hunk 15 above)
Comment (savente93) [reply hydromt/stats/skills.py:14]: Because this docstring contains \ as part of the LaTeX code, so it is necessary for that but not for Others.

File: hydromt/new_feature.py
Comment (savente93) [reply hydromt/new_feature.py:1]: sorry I must have accidentally commited it, I think it's fine to remove. Good catch!

File: hydromt/raster.py (hunk 14 above)
Comment (savente93) [reply hydromt/raster.py:25]: indeed. F401 means no unused imports. I'll add them to the docs.

File: docs/dev/contributing.rst [hunk 20]
```diff
@@ -29,7 +29,7 @@ Checklist pull requests
 If you found a bug or an issue you would like to tackle or contribute to a new development, please make sure do the following steps:
 1. If it does not yet exist, create an issue following the :ref:`issue conventions <Issue conventions>`
 2. Create a new branch where you develop your new code, see also :ref:`Git conventions <Git conventions>`
-3. Run black before committing your changes, see  :ref:`code format <Code format>`. This does only apply for *.py files. For *ipynb files make sure that you have cleared all results.
+3. Make sure all pre-commit hooks pass, see  :ref:`code format <Code format>`. For *ipynb files make sure that you have cleared all results.
```
Comment (hboisgon) [inline docs/dev/contributing.rst:32]:
```suggestion
3. Make sure all pre-commit hooks pass, see  :ref:`code format <Code format>`. For ipynb files make sure that you have cleared all results.
```
The * creates italic that does not stop

File: docs/dev/contributing.rst [hunk 21]
```diff
@@ -185,22 +277,101 @@ You can install pre-commit by running:
```
Comment (hboisgon) [inline docs/dev/contributing.rst:277]: I cant change that line directly but the link does not work (_ missing at the end): and `pre-commit <https://pre-commit.org>`_ to keep everything formatted

File: docs/dev/contributing.rst [hunk 22]
```diff
@@ -185,22 +277,101 @@ You can install pre-commit by running:
...
+
+Now you can push your commit as normal.
+
+From time to time you might see comments like these:
+
+.. code-block:: python
+
+  import rioxarray # noqa: F401
+
+The `noqa` is instructing the linters to igore the specified rule for the line in question. Whenever possible, we try to avoid using these but it's not allways possible. The full list of rules can be found here: `Ruff Rules Seciont <https://beta.ruff.rs/docs/rules/>`__ Some common ones are:
```
Comment (hboisgon) [inline docs/dev/contributing.rst:370]:
```suggestion
The `noqa` is instructing the linters to ignore the specified rule for the line in question. Whenever possible, we try to avoid using these but it's not always possible. The full list of rules can be found here: `Ruff Rules Section <https://beta.ruff.rs/docs/rules/>`_ Some common ones are:
```

File: hydromt/cli/main.py [hunk 23]
```diff
@@ -254,20 +260,25 @@ def update(
...
 
-    \b
-    Update (overwrite!) landuse-landcover based maps in a Wflow model
-    hydromt update wflow /path/to/model_root -c setup_lulcmaps --opt lulc_fn=vito -d /path/to/data_catalog.yml -v
-
-    \b
-    Update Wflow model components outlined in an .ini configuration file and write the model to a directory
-    hydromt update wflow /path/to/model_root -o /path/to/model_out -i /path/to/wflow_config.ini -d /path/to/data_catalog.yml -v
+    Update (overwrite!) landuse-landcover based maps in a Wflow model:
+    hydromt update wflow /path/to/model_root \
```
Comment (hboisgon) [inline hydromt/cli/main.py:271]: Same comment as in build, would be nice if the command line example here and below are in one line

File: hydromt/cli/main.py [hunk 24]
```diff
@@ -334,22 +345,22 @@ def update(
...
     If the existing model contains forcing, they will also be clipped to the new model.
 
     For options to build wflow models see:
 
-    \b
     Example usage to clip a wflow model for a subbasin derived from point coordinates
-    snapped to cells with stream order >= 4
-    hydromt clip wflow /path/to/model_root /path/to/model_destination "{'subbasin': [-7.24, 62.09], 'wflow_streamorder': 4}"
+    snapped to cells with upstream area >= 50 km2
+    hydromt clip wflow /path/to/model_root /path/to/model_destination \
```
Comment (hboisgon) [inline hydromt/cli/main.py:355]: Here as well

File: hydromt/cli/main.py [hunk 25]
```diff
@@ -334,22 +345,22 @@ def update(
...
+        "{'subbasin': [-7.24, 62.09], 'wflow_uparea': 50}"
 
-    \b
     Example usage basin based on ID from model_root basins map
     hydromt clip wflow /path/to/model_root /path/to/model_destination "{'basin': 1}"
 
-    \b
     Example usage basins whose outlets are inside a geometry
-    hydromt clip wflow /path/to/model_root /path/to/model_destination "{'outlet': 'geometry.geojson'}"
+    hydromt clip wflow /path/to/model_root /path/to/model_destination \
```
Comment (hboisgon) [inline hydromt/cli/main.py:362]: Here as well

File: hydromt/models/model_grid.py [hunk 26]
```diff
@@ -151,8 +154,9 @@ def setup_grid_from_raster_reclass(
         rename: Optional[Dict] = dict(),
         **kwargs,
     ) -> List[str]:
-        """
-        HYDROMT CORE METHOD: Add data variable(s) to grid object by reclassifying the data in ``raster_fn`` based on ``reclass_table_fn``.
+        """HYDROMT CORE METHOD: Add data variable(s) to grid object by reclassifying the data in ``raster_fn`` based on ``reclass_table_fn``.
+
+        Reclassifications are based on ``reclass_table_fn``.
```
Comment (hboisgon) [inline hydromt/models/model_grid.py:159]: This one can be removed then

File: hydromt/models/model_mesh.py [hunk 27]
```diff
@@ -112,42 +114,64 @@ def setup_mesh_from_raster_reclass(
         rename: Optional[Dict] = dict(),
         **kwargs,
     ) -> List[str]:
-        """
-        HYDROMT CORE METHOD: Add data variable(s) to mesh object by reclassifying the data in ``raster_fn`` based on ``reclass_table_fn``.
+        """HYDROMT CORE METHOD: Add data variable(s) to mesh object by reclassifying the data in ``raster_fn`` based on ``reclass_table_fn``.
 
-        The reclassified raster data are subsequently interpolated to the mesh using ``resampling_method``.
+        Reclassification is based on `reclass_table_fn`. The reclassified raster data
```
Comment (hboisgon) [inline hydromt/models/model_mesh.py:119]:
```suggestion
        The reclassified raster data
```

File: hydromt/workflows/grid.py [hunk 28]
```diff
@@ -218,12 +224,13 @@ def grid_from_geodataframe(
...
 ) -> xr.Dataset:
-    """
-    Prepares data variable(s) resampled to grid_like object by rasterizing the data from ``gdf``.
-    Several type of rasterization are possible:
-        * "fraction": the fraction of the grid cell covered by the gdf shape is returned.
-        * "area": the area of the grid cell covered by the gdf shape is returned.
-        * "value": the value from the variables columns of gdf are used. If this is used, variables must be specified.
+    """Prepare data variable(s) resampled to grid_like object by rasterizing the data from ``gdf``.
+
+    Rasterizes the data from ``gdf``. Several type of rasterization are possible:
```
Comment (hboisgon) [inline hydromt/workflows/grid.py:229]:
```suggestion
    Several type of rasterization are possible:
```

**Review by hboisgon — APPROVED** [review]
Getting there nice :) A couple more small mistakes but not much. I'll approve for now in advance but would still be good if you generate the docs locally to double-check what you added in contributing.rst. You can see (format) mistakes much more easily there. You know how to do that?

**Comment by savente93** [comment]
Good idea, I did and fixed a small formatting issue. Thanks for the time invested! After the final tests pass, we should be good to go :)
